starky/src/traits.rs

```diff
@@ -10,33 +12,70 @@ pub trait MTNodeType {
     fn as_bn128(self) -> crate::field_bn128::Fr;
 }
 
-pub trait MerkleTree {
-    type MTNode: Copy + std::fmt::Display + Clone + Default + MTNodeType + core::fmt::Debug;
+pub trait MerkleTreeBase {
     type BaseField: Clone
-        + Default
-        + core::fmt::Debug
-        + Into<crate::serializer::Input<Self::MTNode>>;
+    + Default
+    + core::fmt::Debug
+    + Into<crate::serializer::Input<Self::MTNode>>;
+    type MTNode: Copy + std::fmt::Display + Clone + Default + MTNodeType + core::fmt::Debug;
+    type PoseidonType: PoseidonTrait;
+
+    fn merklize_level(&mut self, p_in: usize, n_ops: usize, p_out: usize) -> EigenResult<()>;
+
+    fn do_merklize_level(
+        &self,
+        buff_in: &[Self::MTNode],
+        _st_i: usize,
+        _st_n: usize,
+    ) -> EigenResult<Vec<Self::MTNode>>;
+
+    fn merkle_gen_merkle_proof(&self, idx: usize, offset: usize, n: usize) -> Vec<Vec<Self::BaseField>>;
+
+    fn merkle_calculate_root_from_proof(
+        &self,
+        mp: &Vec<Vec<Self::BaseField>>,
+        idx: usize,
+        value: &Self::MTNode,
+        offset: usize,
+    ) -> EigenResult<Self::MTNode>;
+
+    fn calculate_root_from_group_proof(
+        &self,
+        mp: &Vec<Vec<Self::BaseField>>,
+        idx: usize,
+        vals: &Vec<FGL>,
+    ) -> EigenResult<Self::MTNode>;
+}
+
+pub trait MerkleTree: MerkleTreeBase {
```

Comment:
We may not have to define a MerkleTreeBase?  We can add new methods, like merkle_gen_merkle_proof, to this traits?